### ISSUE TITLE
restrict host normalization to http/https

### ISSUE THREAD

**Issue by nateprewitt** [shazow/urllib3#1080]
This is an offshoot from further discussion in kennethreitz/requests#3738 beyond the scope of the PR.

#911 introduced normalizing both the host and scheme to lowercase which is valid for http URIs. Requests however was supporting the passing of percent-encoded file paths with the http+unix:// scheme which is now broken as of Requests 2.12.0 (urllib3 1.17+). @Lukasa [suggested](https://github.com/kennethreitz/requests/pull/3738#issuecomment-268883231) we could restrict host normalization to a specific set of schemes to avoid breaking case-dependent URLs.

What we want to do is only run [this check](https://github.com/shazow/urllib3/blob/4851f6d5e23851b497cd30cac89ebcabf5b8853c/urllib3/util/url.py#L24-L25) when there is a specified scheme and it is inside of a predefined set of normalizable schemes (http://, https://, ?).

@Lukasa may have more details on how he'd like this being implemented, but this would likely be a good issue for someone looking to contribute to urllib3.

**Comment by Lukasa** [comment]
So, let's step back and consider the whole problem space.

The question boils down to: is the host part of a URI case sensitive or not? RFC 3986 says it is, but also says in the same breath that they shouldn't necessarily be unilaterally downcased:

> Although host is case-insensitive, producers and normalizers should use lowercase for registered names and hexadecimal addresses for the sake of uniformity, while only using uppercase letters for percent-encodings.

Ultimately, however, RFC 3986 section 6.2.2.1 (Case Normalization) says:

> When a URI uses components of the generic syntax, the component syntax equivalence rules always apply; namely, that the scheme and host are case-insensitive and therefore should be normalized to lowercase.  For example, the URI <HTTP://www.EXAMPLE.com/> is equivalent to <http://www.example.com/>.

Reading between the lines, this implies that a scheme may choose to not use a component of the generic syntax, or may define its own rules. That means that we cannot assume that all schemes will case-normalize their host. Of course, what it *also* means is that not all schemes can sensibly be parsed by urllib3 *at all*.

The conservative change here would be to say that we should only do anything that looks like a case-normalization step for a specific set of schemes[^1] that we know employ case normalization. In the case of urllib3, that would be 'http' and 'https', though I should note that there are a few other URI schemes that would allow it ('ftp' being the first example that leaps to mind).

I would advocate for this conservative change to be made in urllib3 for the moment, because it is the smallest and most reasonable change that will resolve the problem in Requests. However, all of this continues to reveal problems with the loosey-goosey way that Requests chooses to work with URLs, and is a continuing punishment for our previous refusal to be strict. urllib3 has a reasonable case to argue that the `URL` class is fundamentally about *HTTP* only, and all non-HTTP schemes should take a long walk off a short pier. It would also be reasonable for urllib3 to turn `host` into a property that does the case transformation so that Requests can get at the un-case-normalized version if it feels it needs to.

Ultimately, though, all of this is gross. Requests' extensibility here is causing us trouble. =(

[^1]: I should note while I'm here that we can always safely case-normalize the scheme because the scheme part of a URI must always be generic. That's not a concern.

**Comment by tiran** [comment]
Request should not perform normalization by itself. Instead case normalization of host names should be handled by URT46 mapping in IDNA. Although normalization may like trivial at first, it is far from trivial. In unicode normalization is not necessarily bijective or reversible. https://labs.spotify.com/2013/06/18/creative-usernames/ is a fun example. AFAIK IDNA imposes further restrictions on normalization, e.g. lower casing after NFC normalization for a limited subset of unicode.

The real question should be: Should urllib3 perform IDNA encoding on a string or not? Which brings us back to the initial topic of the issue -- restrict host normalization to http and https. :)

@Lukasa sorry, I forgot to include test case for upper case chars in the host name.

```
$ git diff
diff --git a/tests/test_requests.py b/tests/test_requests.py
index b0cb7d6..9aaf56a 100755
--- a/tests/test_requests.py
+++ b/tests/test_requests.py
@@ -2232,10 +2232,18 @@ class TestPreparingURLs(object):
                 u"http+unix://%2fvar%2frun%2fsocket/path",
             ),
             (
+                b"http+unix://%2Fvar%2Frun%2FSOCKET/path",
+                u"http+unix://%2fvar%2frun%2fSOCKET/path",
+            ),
+            (
                 u"http+unix://%2Fvar%2Frun%2Fsocket/path",
                 u"http+unix://%2fvar%2frun%2fsocket/path",
             ),
             (
+                u"http+unix://%2Fvar%2Frun%2FSOCKET/path",
+                u"http+unix://%2fvar%2frun%2fSOCKET/path",
+            ),
+            (
                 b"mailto:user@example.org",
                 u"mailto:user@example.org",
             ),

```

```
>       assert p.url == expected
E       assert 'http+unix://...2fsocket/path' == 'http+unix://%...2fSOCKET/path'
E         - http+unix://%2fvar%2frun%2fsocket/path
E         ?                            ^^^^^^
E         + http+unix://%2fvar%2frun%2fSOCKET/path
E         ?  
```

**Comment by Lukasa** [comment]
Does URT46 normalisation apply to ASCII characters, both for the idna module and in the abstract? If it does we may be able to handle this that way instead.

**Comment by tiran** [comment]
Cory and I discuss the matter on IRC. For the record, the idna package performs both case normalization and NFC normalization for ASCII and non-ASCII chars

```
>>> idna.encode(u'Straße.DE', uts46=True)
'xn--strae-oqa.de'
>>> idna.encode(u'Street.COM', uts46=True)
'street.com'
>>> idna.encode(u'ba\N{COMBINING DIAERESIS}cker', uts46=True)
'xn--bcker-gra'
>>> idna.encode(u'bäcker', uts46=True)
'xn--bcker-gra'
>>> idna.encode(u'bÄcker', uts46=True)
'xn--bcker-gra'
```

**Comment by Lukasa** [comment]
So this would mean that our ideal policy would be:

1. If the URI has a scheme that is not http/https, leave it alone.
2. Otherwise, apply IDNA.
3. If that fails, and the hostname is entirely ASCII, then lowercase the hostname.

Unfortunately, that is moderately annoying, because urllib3 does not rely on having the idna module present. If we come to depend on it, then requests needs to start rewriting urllib3 imports. So I'm not sure what to do there.

Perhaps, for now, it's best to just do steps 1 and 3.

**Comment by tiran** [comment]
I agree, it would be nice to have IDNA 2008 / UTS46 in Python's standard library *wink wink nudge nudge*.

**Comment by nateprewitt** [comment]
So I think we're in agreement that, for now, we only need to bound the host check with an additional `and scheme in ('http', 'https')` and a test for a scheme(s) outside of that set.

It looks like the state of IDNA in urllib3 is currently in limbo for v2, so we can probably address that later if needed.

**Comment by Lukasa** [comment]
Yeah, urllib3 doesn't really do IDNA very much but it often kinda sorta worked because httplib kinda sorta did IDNA but not very well.

We really need to reconcile that behaviour inside urllib3: specifically, we need to decide either that we think that all hostnames are ASCII and its the user's responsibility to handle IDNA, or we need to decide to plumb IDNA support into the codebase in a sensible way so that all layers are consistently working on either internationalized or non-internationalized domain names.

**Comment by lslebodn** [comment]
Is there any PR open for this issue?

**Comment by nateprewitt** [comment]
Not currently @Islebodn, I think it's fair game if you'd like to pick it up. Otherwise, I'll throw something together later this week.

**Comment by lslebodn** [comment]
Just a tiny question. Is there a plan to release new version of the package or should I ask to backport the patch to debian

**Comment by Lukasa** [comment]
urllib3 and requests will ship new versions, but I wasn't planning to shove a new release out in the near future, so feel free to get debian to backport if you'd like.

**Comment by lslebodn** [comment]
Thank you very much for info.

**Comment by nateprewitt** [comment]
Resolved with #1089.
